buildvertical.targets

```diff
@@ -51,35 +58,57 @@
+          Returns="$(_AllBuildConfigurations)">
+     <PropertyGroup>
+       <_AllBuildConfigurations>$(BuildConfigurations)</_AllBuildConfigurations>
+       <_AllBuildConfigurations Condition="'$(BuildConfigurations)' == ''">$(BuildConfiguration)</_AllBuildConfigurations>
+     </PropertyGroup>
   </Target>
 
   <!-- Runs in a leaf project (csproj) to determine best configuration for ProjectReferences -->
   <Target Name="AnnotateProjectReference"
-          BeforeTargets="BeforeResolveReferences">
+          BeforeTargets="BeforeResolveReferences"
+          Condition="'@(ProjectReference)' != ''"
+          Inputs="%(ProjectReference.Identity)"
+          Outputs="unused">
 
-    <!-- find the best configuration for each project, projects that do not have an 
-        applicable configuration will fail the build.  -->
-    <MSBuild Targets="FindBestConfigurationForProjectReference"
+    <!-- Get all configurations for each ProjectReference.  We use this 
+         rather than the FindBestConfiguration target since we need to 
+         choose the best configuration for the referencing project's 
+         configuration, not the BuildConfiguration.  -->
+    <MSBuild Targets="GetBuildConfigurations"
              Projects="@(ProjectReference)">
-      <Output TaskParameter="TargetOutputs" ItemName="_ProjectReferenceBestConfigurations" />
+      <Output TaskParameter="TargetOutputs" PropertyName="_projectReferenceBuildConfigurations" />
     </MSBuild>
 
+    <!-- Find the best configuration for the current Project's Configuration -->
+    <FindBestConfiguration Condition="'$(_projectReferenceConfiguration)' == ''"
+                           Properties="@(Property)"
+                           PropertyValues="@(PropertyValue)"
+                           BuildConfigurations="$(_ProjectReferenceBuildConfigurations)"
+                           BuildConfiguration="$(Configuration)">
+      <Output TaskParameter="BestConfiguration" PropertyName="_projectReferenceConfiguration" />
+    </FindBestConfiguration>
+
+    <Error Condition="'$(_projectReferenceConfiguration)' == ''" Text="Could not find a configuration for ProjectReference '@(ProjectReference)' from configurations '$(_projectReferenceBuildConfigurations)' when building '$(MSBuildProjectName)' for configuration '$(Configuration)' ." />
+
     <ItemGroup>
-      <!-- transform back to ProjectReference -->
-      <ProjectReferenceWithConfiguration Include="@(_ProjectReferenceBestConfigurations->'%(OriginalItemSpec)')">
-        <AdditionalProperties>Configuration=%(Identity);%(_ProjectReferenceBestConfigurations.AdditionalProperties)</AdditionalProperties>
-      </ProjectReferenceWithConfiguration>
+      <ProjectReference>
+        <AdditionalProperties>Configuration=$(_projectReferenceConfiguration);%(ProjectReference.AdditionalProperties)</AdditionalProperties>
+      </ProjectReference>
     </ItemGroup>
+  </Target>
 
+  <Target Name="BuildAll">
```

Comment:
Yes it works from an execution perspective.  I'm mainly concerned with making it work from CSProj but I could add another property to make it work for traversal too.  We still have a lot of projects to fix up to make the whole repo work, but this does work for an individual project.